src/art/art.c

```diff
@@ -1448,118 +1633,301 @@ static void art_node_printf(const art_node_t *node, uint8_t depth) {
+            break;
+        }
+        default: {
+            assert(false);
+            return 0;
+        }
+    }
+    art->first_free[typecode] = art_next_free(art, typecode, to + 1);
+    return art_to_ref(to, typecode);
+}
+
+/**
+ * Shrinks all node arrays to `first_free`. Assumes all indices after
+ * `first_free` are unused.
+ */
+static size_t art_shrink_node_arrays(art_t *art) {
+    size_t freed = 0;
+    if (art->first_free[CROARING_ART_LEAF_TYPE] <
+        art->capacities[CROARING_ART_LEAF_TYPE]) {
+        size_t new_capacity = art->first_free[CROARING_ART_LEAF_TYPE];
+        art->leaves =
+            roaring_realloc(art->leaves, new_capacity * sizeof(art_leaf_t));
+        freed += art->capacities[CROARING_ART_LEAF_TYPE] - new_capacity;
+        art->capacities[CROARING_ART_LEAF_TYPE] = new_capacity;
+    }
+    if (art->first_free[CROARING_ART_NODE4_TYPE] <
+        art->capacities[CROARING_ART_NODE4_TYPE]) {
+        size_t new_capacity = art->first_free[CROARING_ART_NODE4_TYPE];
+        art->node4s =
+            roaring_realloc(art->node4s, new_capacity * sizeof(art_node4_t));
+        freed += art->capacities[CROARING_ART_NODE4_TYPE] - new_capacity;
+        art->capacities[CROARING_ART_NODE4_TYPE] = new_capacity;
+    }
+    if (art->first_free[CROARING_ART_NODE16_TYPE] <
+        art->capacities[CROARING_ART_NODE16_TYPE]) {
+        size_t new_capacity = art->first_free[CROARING_ART_NODE16_TYPE];
+        art->node16s =
+            roaring_realloc(art->node16s, new_capacity * sizeof(art_node16_t));
+        freed += art->capacities[CROARING_ART_NODE16_TYPE] - new_capacity;
+        art->capacities[CROARING_ART_NODE16_TYPE] = new_capacity;
+    }
+    if (art->first_free[CROARING_ART_NODE48_TYPE] <
+        art->capacities[CROARING_ART_NODE48_TYPE]) {
+        size_t new_capacity = art->first_free[CROARING_ART_NODE48_TYPE];
+        art->node48s =
+            roaring_realloc(art->node48s, new_capacity * sizeof(art_node48_t));
+        freed += art->capacities[CROARING_ART_NODE48_TYPE] - new_capacity;
+        art->capacities[CROARING_ART_NODE48_TYPE] = new_capacity;
+    }
+    if (art->first_free[CROARING_ART_NODE256_TYPE] <
+        art->capacities[CROARING_ART_NODE256_TYPE]) {
+        size_t new_capacity = art->first_free[CROARING_ART_NODE256_TYPE];
+        art->node256s = roaring_realloc(art->node256s,
+                                        new_capacity * sizeof(art_node256_t));
+        freed += art->capacities[CROARING_ART_NODE256_TYPE] - new_capacity;
+        art->capacities[CROARING_ART_NODE256_TYPE] = new_capacity;
+    }
+    return freed;
+}
+
+/**
+ * Traverses the ART, moving nodes to earlier free indices and modifying their
+ * references along the way.
+ */
+static void art_shrink_at(art_t *art, art_ref_t ref) {
+    if (art_is_leaf(ref)) {
         return;
     }
-    art->root = art_insert_at(art->root, key, 0, leaf);
+    switch (art_ref_typecode(ref)) {
```

Comment:
`art_shrink_at` currently unconditionally visits all nodes. I think it might be worth it to check if any node arrays _need_ shrinking: If all `first_free`s are equal to their capacity, we can just immediately return, I believe.